flink-java/src/main/java/org/apache/flink/api/java/utils/PythonDependencyManager.java

```diff
@@ -0,0 +1,282 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.api.java.utils;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.common.cache.DistributedCache;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.configuration.ConfigOption;
+import org.apache.flink.configuration.ConfigOptions;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.configuration.PythonOptions;
+import org.apache.flink.configuration.ReadableConfig;
+import org.apache.flink.util.Preconditions;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+import static org.apache.flink.configuration.PythonOptions.PYTHON_CLIENT_EXECUTABLE;
+import static org.apache.flink.configuration.PythonOptions.PYTHON_EXECUTABLE;
+
+/**
+ * Utility class for Python dependency management. The dependencies will be registered at the distributed
+ * cache.
+ */
+@Internal
+public class PythonDependencyManager {
+
+	public static final String PYTHON_FILE_PREFIX = "python_file";
+	public static final String PYTHON_REQUIREMENTS_FILE_PREFIX = "python_requirements_file";
+	public static final String PYTHON_REQUIREMENTS_CACHE_PREFIX = "python_requirements_cache";
+	public static final String PYTHON_ARCHIVE_PREFIX = "python_archive";
+	public static final String CLUSTER = "cluster";
+	public static final String FILE = "file";
+	public static final String CACHE = "cache";
+	public static final String FILE_DELIMITER = ",";
+	public static final String PARAM_DELIMITER = "#";
+
+	// Metadata of Python dependencies.
+
+	public static final ConfigOption<Map<String, String>> PYTHON_FILES =
+		ConfigOptions.key("python.internal.files-key-map").mapType().noDefaultValue();
+	public static final ConfigOption<Map<String, String>> PYTHON_REQUIREMENTS_FILE =
+		ConfigOptions.key("python.internal.requirements-file-key").mapType().noDefaultValue();
+	public static final ConfigOption<Map<String, String>> PYTHON_ARCHIVES =
+		ConfigOptions.key("python.internal.archives-key-map").mapType().noDefaultValue();
+
+	private final Configuration config;
+	private final List<Tuple2<String, DistributedCache.DistributedCacheEntry>> cachedFiles;
+	private int counter = -1;
+
+	public PythonDependencyManager(
+		Configuration config,
+		List<Tuple2<String, DistributedCache.DistributedCacheEntry>> cachedFiles) {
+		this.config = Preconditions.checkNotNull(config);
+		this.cachedFiles = cachedFiles;
+	}
+
+	/**
+	 * Adds a Python dependency which could be .py files, Python packages(.zip, .egg etc.) or
+	 * local directories. The dependencies will be added to the PYTHONPATH of the Python UDF worker
+	 * and the local Py4J python client.
+	 *
+	 * @param filePath The path of the Python dependency.
+	 */
+	public void addPythonFile(String filePath) {
+		Preconditions.checkNotNull(filePath);
+		if (fileAlreadyRegistered(filePath)) {
```

Comment:
Should we throw an exception if this case occurs as usually this is not expected?

Reply:
This case is expected because the `configure()` method will be called when the 
`translate()` method of the Table API planners is called. As the `translate()` can be called many times in interactive programing, a python file will be registered multiple times.